model-optimizer/extensions/back/MarkNodesWithShapeValues.py

```diff
@@ -0,0 +1,88 @@
+from mo.back.replacement import BackReplacementPattern
+from mo.graph.graph import Graph
+
+
+class MarkNodesWithShapeValues(BackReplacementPattern):
+    """
+    This replacer marks op nodes in ShapeOf subgraphs with 'returns_shape_value' bool attribute and
+    data nodes of float32 constants with 'correct_data_type' attribute.
+    So that float Consts and Cast_to float will be kept in FP32 even if cmd_param --data_type=FP16 was specified.
+
+    This is needed to enable conversion to FP16 even if values in ShapeOf subgraphs exceed max(float16)
+    or because of FP16 lower precession shape inference is incorrect on some nodes (e.g. if Interpolate in scales mode
+    accepts values from ShapeOf subgraph).
+
+    This transformation must be run after shape inference and after all transformations that insert/modify
+    Cast nodes in ShapeOf subgraphs therefore it's placed at the end of the back phase.
+    """
+    enabled = True
+    graph_condition = [lambda graph: graph.graph['cmd_params'].data_type == 'FP16']
+
+    def run_after(self):
+        from extensions.back.pass_separator import BackFinish
+        return [BackFinish]
+
+    def run_before(self):
+        return []
+
+    @staticmethod
+    def get_operations_with_shape_inputs():
+        return {
+            'Interpolate': [1, 2],  # sizes, scales inputs
+            'Reshape': [1],  # shape
+            'Broadcast': [1],  # target_shape
+            'ConvBackPropData ': [2],  # output_shape
+            'GroupConvolutionBackpropData ': [2],  # output_shape
+            'BatchToSpace': [1, 2, 3],  # block_shape, crops_begin, crops_end
+            'SpaceToBatch': [1, 2, 3],  # block_shape, pads_begin, pads_end
+            'StridedSlice': [1, 2, 3],  # begin, end, strides
+            'VariadicSplit': [2],  # split_lengths
+            'Tile': [1],  # repeats input
+            'TopK': [1],  # K input
+            'Pad': [1, 2],  # pads_begin, pads_end
+            'Range': [0, 1, 2],  # start, stop, step inputs
+            'OneHot': [1],  # depth input
+        }
+
+    def find_and_replace_pattern(self, graph: Graph):
+        shape_input_ops_map = self.get_operations_with_shape_inputs()
+        nodes_with_shape_inputs = []
+        for op_type in shape_input_ops_map:
+            nodes_with_shape_inputs.extend(graph.get_op_nodes(type=op_type))
+
+        start_nodes = []
+        for node in nodes_with_shape_inputs:
+            start_ports = [x for x in shape_input_ops_map[node.type] if not node.in_port(x).disconnected()]
+            start_nodes.extend([node.in_port(port_idx).get_source().node for port_idx in start_ports])
+
+        condition = lambda node: node.soft_get('type') != 'ShapeOf'
+        nodes_with_shape_values = MarkSubGraphsWithCorrectLayout.bfs(start_nodes, set(), condition, forward=False)
+        for node in nodes_with_shape_values:
+            node['returns_shape_value'] = True
```

Comment:
Probably, we need to set  `'returns_shape_value'` to `False` for all nodes in the graph first. Like this:
`nx.set_node_attributes(G=graph, name='returns_shape_value', values=False)`

Reply:
We have wonderful methods like `has_and_set` or `soft_get` and use them quite often. I don't see any necessity for setting this attribute for every singe node in the graph.